python/pyiceberg/schema.py

```diff
@@ -638,3 +724,61 @@ def map(self, map_type: MapType, key_result: int, value_result: int) -> int:
 
     def primitive(self, primitive: PrimitiveType) -> int:
         return 0
+
+
+def assign_fresh_schema_ids(schema: Schema) -> Schema:
+    """Traverses the schema, and sets new IDs"""
+    schema_struct = pre_order_visit(schema.as_struct(), _SetFreshIDs())
+
+    fresh_identifier_field_ids = []
+    new_schema = Schema(*schema_struct.fields)
+    for field_id in schema.identifier_field_ids:
+        original_field_name = schema.find_column_name(field_id)
+        if original_field_name is None:
+            raise ValueError(f"Could not find field: {field_id}")
+        fresh_field = new_schema.find_field(original_field_name)
+        if fresh_field is None:
+            raise ValueError(f"Could not lookup field in new schema: {original_field_name}")
+        fresh_identifier_field_ids.append(fresh_field.field_id)
+
+    return new_schema.copy(update={"identifier_field_ids": fresh_identifier_field_ids})
+
+
+class _SetFreshIDs(PreOrderSchemaVisitor[IcebergType]):
+    """Traverses the schema and assigns monotonically increasing ids"""
+
+    counter: itertools.count
+
+    def __init__(self, start: int = 1) -> None:
+        self.counter = itertools.count(start)
+
+    def _get_and_increment(self) -> int:
+        return next(self.counter)
+
+    def schema(self, schema: Schema, struct_result: Callable[[], StructType]) -> Schema:
+        return Schema(*struct_result().fields, identifier_field_ids=schema.identifier_field_ids)
```

Comment:
Shouldn't this re-map the identifier field IDs since it is returning a new schema?

Reply:
Yes, we do that in the function itself:
```python
def assign_fresh_schema_ids(schema: Schema) -> Schema:
    """Traverses the schema, and sets new IDs"""
    schema_struct = pre_order_visit(schema.as_struct(), _SetFreshIDs())

    fresh_identifier_field_ids = []
    new_schema = Schema(*schema_struct.fields)
    for field_id in schema.identifier_field_ids:
        original_field_name = schema.find_column_name(field_id)
        if original_field_name is None:
            raise ValueError(f"Could not find field: {field_id}")
        fresh_field = new_schema.find_field(original_field_name)
        if fresh_field is None:
            raise ValueError(f"Could not lookup field in new schema: {original_field_name}")
        fresh_identifier_field_ids.append(fresh_field.field_id)

    return new_schema.copy(update={"identifier_field_ids": fresh_identifier_field_ids})
```
This is because we first want to know all the IDs

Reply:
Ah, I've refactored this because we need to build a map anyway 👍🏻